### DISCUSSION TITLE
Context Deadline Exceeded?

### DISCUSSION THREAD

**Discussion by Mashiane** [pocketbase/pocketbase#5847] [Q&A]
Hi

I need some advise. My collection has 17K rows however at times its able to return my query records and sometimes it gives a context deadline exceeded. Ive googled what this means and its tricky.

I am attaching my collection here
[pb_schema (2).zip](https://github.com/user-attachments/files/17719024/pb_schema.2.zip)

This is one of the log examples that resulted in this also.

![image](https://github.com/user-attachments/assets/5eeac8dd-ab36-488f-9a9b-3a1344e7ab9d)

Im using js sdk for the queries in some instances.

Thank you in advance.

**Comment by ganigeorgiev** [comment]
The error most likely indicates that your query takes too long to complete. I don't have a sandboxed environment prepared at the moment to download random zip files so I can't test or view your collections configuration _(try to attach it either as plain json file or paste it directly as text wrapped in 3backticks)_.

To verify that this is the case you can try increasing the default query time limit with the `--queryTimeout=120` flag (the value is in seconds).

Why the query takes so long it is difficult to say. The number of records definitely have effect but it is far from the only reason that could cause it (in Presentator I have a collection with almost 300k records and it loads and filters fine).

Common culprits for slow queries are:

- large number of columns or raw data within a single record
- too many joins (e.g. dot-notation access in API rule or filter)
- lack of indexes
- too many and incorrect indexes (see https://github.com/pocketbase/pocketbase/discussions/3859#discussioncomment-7744045)

If you have difficulties finding which indexes you need:
1. start the executable with `--dev`
2. sent your request
3. copy the SQL query from the terminal related to your request (_you may have to replace `[[]]` and 
 `{{}}` placeholders; this is done automatically with v0.23_)
4. run `sqlite3 pb_data/data.db`
5. run `EXPLAIN QUERY PLAN yourFinalQuery;`
   Or alternatively you can also try the `.expert` command to suggest you the indexes:
   ```
   sqlite> .expert
   sqlite> yourFinalQuery;
   ```

**Comment by lumi-mustafa-hanif** [comment]
<img width="1325" alt="Screenshot 2025-03-13 at 12 33 29 PM" src="https://github.com/user-attachments/assets/a4296b44-37a1-4052-97da-63491511cbfd" />
This is my hetzner graph on a low activity. I have the best instance which is cax31.

I am planning a way to show you and give you access to my entire stack, one immediate issue I can give exact instructions are batch delete.

- Have a table with 50k rows.
- From nodeJS code, create a batch service call to delete 100 records, and after those are done 100 more until table is empty
- It will take a long time to delete, CPU will go to maximum usage much after the deletion.

**Reply by ganigeorgiev** [reply]
@mustafa-hanif Again as mentioned in your original post https://github.com/pocketbase/pocketbase/discussions/6579, you are not providing details about your environment, what PocketBase version you are using and how to reproduce the issue in general.

Are you even able to reproduce it locally? Are you sure that the issue is not caused by the Node.js code and not by PocketBase? Did you compare it against bulk deleting the records from the Admin UI?

Do the records you are deleting have files? If you are using S3 then the slow deletion and high CPU utilization could be from slow requests to the S3 server.

If by "create a batch service" you mean the [batch Web API](https://pocketbase.io/docs/api-records/#batch-createupdateupsertdelete-records) then it is possible to be slower because there are a lot of checks and copying involved and if the goal is to bulk delete records then you'll be able to achieve better performance if you send multiple concurrent individual delete requests.

Also note that PocketBase v0.24+ added significant speed up to delete operations in general, see the benchmarks - https://github.com/pocketbase/benchmarks/blob/master/results/hetzner_cax11.md#deleting-records.

I know that you are frustrated but understand that no one here has any context about your project and how it works. Please try to create a **minimal** reproducible repo and I'll try to have a look at it but just saying _"create a batch service call to delete 100 records"_ doesn't really help much with reproducing the issue.

**Reply by ganigeorgiev** [reply]
I'm marking this as off-topic to prevent confusion since the above CPU utilization issue is not related to the original "context deadline exceeded" post.

@mustafa-hanif Please don't post in multiple places and try to do a followup in your original issue https://github.com/pocketbase/pocketbase/discussions/6579.
